text/0000-destructor-guarantees.md

```diff
@@ -0,0 +1,176 @@
+
+### This Proposal Will Not:
+
+ - Prevent *all* destructors from not running.
+ - Prevent destructors from not running in the face of process aborts.
+ - Prevent destructors from not running for `'static` data.
+ - Prevent `Rc` or `Arc` cycles, or ban non-`'static` data from them.
+
+### This Proposal Will:
+
+ - Allow types to manually opt-in to not being leaked.
+ - Allow leaks of `'static` data in a large variety of ways, even if they do
+   not implement `Leak`.
+ - Prevent types which have opted-in to not being leaked from being used in
+   contexts where they might leak.
+ - Prevent panics in destructors from causing leaks or other memory unsafety.
+ - Allow `Rc` and `Arc` cycles, including when they contain non-`'static` data.
+
+# Detailed design
+
+Introduce a new default trait to `core`, `Leak`:
+
+```rust
+pub unsafe trait Leak { }
+unsafe impl Leak for .. { }
+```
+
+Change several `std` APIs to adjust for the guarantees now provided to types
+which do not implement `Leak`:
+
+ - `impl<'a, T> !Leak for ::std::thread::JoinGuard<'a, T> { }`
+   - If we gain some form of specialization, implement `Leak` for `JoinGuard<'static, T>`
+ - implement `!Leak` for the guards returned by collection `drain` operations
+ - Add a `Leak` bound to the type parameter of:
+   - `rc::Rc`
+   - `sync::Arc`
+   - `sync::mpsc::Sender` (possibly not, see unresolved questions)
+   - Possibly other APIs. Please point any others out if you think of them.
+
+_Cause all panics in destructors to immediately abort the process._
+
+Add a new function, `leak`, to `std::mem`:
+
+```rust
+/// Moves a value out of scope without running its destructor.
+///
+/// `leak` is safe since the data must implement `Leak` and therefore it is
+/// safe to avoid its destructor.
+pub fn leak<T: Leak>(x: T) {
+    unsafe { forget(x) }
+}
+```
+
+There are no changes to `mem::forget`, which remains an `unsafe` way to bypass
+this system.
+
+## When do you need a Leak bound?
+
+If your API can delay or prevent the destructor of non-`'static` data past the
+lifetime of that data, you need a `Leak` bound. If your API requires `'static`
+already you do not need a `Leak` bound.
+
+# Drawbacks
+
+Introduces additional complexity in the form of the `Leak` trait, which, like
+`Reflect`, applies to nearly all types and therefore carries very little
+information.
+
+It is possible there are ways to leak arbitrary, non-`'static`, data in the
+safe subset of rust (including std) that we may not catch by 1.0, which would
+force us to make a breaking fix to those APIs to fix soundness holes if they
+are discovered post-1.0 (by adding a `Leak` bound where there was none
+previously).
+
+If a way to leak in the entirely safe subset of rust + `mem::swap`, then this
+proposal is basically dead on arrival. We should investigate if this is
+possible, but the author's slightly educated guess is that it is not possible.
+
```

Comment:
It may be worth also mentioning that this would likely be a large breaking change due to the addition of bounds on the type parameters for the primitives listed above. (and the short time frame that would have to be reconciled in).